cli/azd/pkg/pipeline/pipeline_manager.go

```diff
@@ -196,50 +196,48 @@ func (pm *PipelineManager) Configure(ctx context.Context) (result *PipelineConfi
 	}
 
 	// If an explicit client id was specified but not found then fail
-	if application == nil && lookupKind == lookupKindPrincipalId {
+	if servicePrincipal == nil && lookupKind == lookupKindPrincipalId {
 		return nil, fmt.Errorf(
 			"service principal with client id '%s' specified in '--principal-id' parameter was not found",
 			pm.args.PipelineServicePrincipalId,
 		)
 	}
 
 	// If an explicit client id was specified but not found then fail
-	if application == nil && lookupKind == lookupKindEnvironmentVariable {
+	if servicePrincipal == nil && lookupKind == lookupKindEnvironmentVariable {
 		return nil, fmt.Errorf(
 			"service principal with client id '%s' specified in environment variable '%s' was not found",
 			envClientId,
 			AzurePipelineClientIdEnvVarName,
 		)
 	}
 
-	if application == nil {
+	if servicePrincipal == nil {
 		displayMsg = fmt.Sprintf("Creating service principal %s", applicationName)
 	} else {
-		displayMsg = fmt.Sprintf("Updating service principal %s (%s)", application.DisplayName, *application.AppId)
+		displayMsg = fmt.Sprintf("Updating service principal %s (%s)", servicePrincipal.DisplayName, servicePrincipal.AppId)
 	}
 
 	pm.console.ShowSpinner(ctx, displayMsg, input.Step)
-	clientId, credentials, err := pm.adService.CreateOrUpdateServicePrincipal(
+	servicePrincipal, err = pm.adService.CreateOrUpdateServicePrincipal(
```

Comment:
I wonder if the `pipeline_manager` should still be the one who either creates a secret-password for the ServicePrincipal or set a FederatedCredential. 

Is there a reason why we want to delegate that responsibility to the pipeline provider? Are we having a different implementation, for example, to create a secret-password for Github and Azdo?

Can the pipeline_manager:
- Create the SP and the client-secret and just pass it to the gh or azdo provider to set it? (when client-secret is selected)
- Create the SP and set up the FederatedCredential, then pass a nil for the credential to gh, so it ignores the credential setting.
- Display a warning when FedCredential + Azdo and fallback to SP+client-secret+Azdo
- Display warning when FedCred + Terraform and fallback to SP+client-secret?  .. (we can support Terraform + FedCred on Github, but it requires us to set both, FedCred for azd-auth and also client-secret for terraform-auth..  so, we need to display warning and decide if we fallback to client-secret or we tell customer that azd will use both FedCred and client-secret)

Reply:
The reason for delegating the authentication type of the service principal is because the pipeline manager does not know what credential types are supported by each CI provider.  For example, github supports federated and client-credentials whereas azdo only support client credentials.

The primary bug that this PR is fixing is that we shouldn't be creating client-credentials if it wasn't asked for.

Reply:
Can we create a new method for CI-Provider interface to return the list of supported-Credential-Types?

I like the idea about the PipelineManager not implementing the logic around what's supported by each CI-Provider, but I don't think a CI-Provider should know anything around creating credentials, as that seems outside of its CI-scope.  

So, I'm wondering about a flow like:

1. PipelineManager resolves and creates the CI-Provider
2. Then, PipelineManager resolves the requested credentials type to use and ask the CI-Provider for the list of type it can support.
3. Then, PipelineManager will return an error if the requested type is not supported by the CI-Provider. Or, if it is supported, the PipelineManager will create the SP and the requested credential type.
4. Finally, the PipelineManager calls the CI-Provider to configure the CI pipeline using the SP+Credential+CredentialType.

This makes a CI-Provider to only care about defining the credential types it knows how to apply and for implementing the way each credential type is applied.  Without knowing about how the credentials are created.  This would follow what's modeled on the real-world, as we don't go to Github to create the secret for a SP, we only go to GitHub/Azdo to setup an existing SP+credential

Reply:
A couple points to note - thinks like client credentials with client secret are very generic and work across most providers.  

However, the federated credential is a different story. The configuration of the federated credential is very specific to the supported provider.  It requires a well known issuer and predefined subject format.  Because of things like this it feels more valid to continue having the CI provider create the credential with the correct configuration.

Reply:
But a CI-Provider shouldn't know how to create a fedCredential for Azure.  
We can make the CI-Provider interface to have a method like `FedCredentialOptions() fedCredOptions` and have the provider to return the required settings for creating a fedCredential.  This will follow the real life scenario where the Azure FedCred is created the same way for any particular party... we just need some input settings for the party.